keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,249 @@
+  - "@lavalamp"
+creation-date: 2018-08-06
+last-updated: 2018-09-25
+status: provisional
+---
+
+# Migrating API objects to latest storage version
+
+## Table of Contents
+
+   * [Migrating API objects to latest storage version](#migrating-api-objects-to-latest-storage-version)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+      * [Proposal](#proposal)
+         * [Alpha workflow](#alpha-workflow)
+         * [Alpha API](#alpha-api)
+         * [Failure recovery](#failure-recovery)
+         * [Beta workflow - Automation](#beta-workflow---automation)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Alternatives](#alternatives)
+         * [update-storage-objects.sh](#update-storage-objectssh)
+
+## Summary
+
+We propose a solution to migrate the stored API objects in Kubernetes clusters.
+In 2018 Q4, we will deliver a tool of alpha quality. The tool extends and
+improves based on the [oc adm migrate storage][] command.  We will integrate the
+storage migration into the Kubernetes upgrade process in 2019 Q1.  We will make
+the migration automatically triggered in 2019.
+
+[oc adm migrate storage]:https://www.mankier.com/1/oc-adm-migrate-storage
+
+## Motivation
+
+"Today it is possible to create API objects (e.g., HPAs) in one version of
+Kubernetes, go through multiple upgrade cycles without touching those objects,
+and eventually arrive at a version of Kubernetes that can’t interpret the stored
+resource and crashes. See k8s.io/pr/52185."[1][]. We propose a solution to the
+problem.
+
+[1]:https://docs.google.com/document/d/1eoS1K40HLMl4zUyw5pnC05dEF3mzFLp5TPEEt4PFvsM
+
+### Goals
+
+A successful storage version migration tool must:
+* work for Kubernetes built-in APIs, custom resources (CR), and aggregated APIs.
+* do not add burden to cluster administrators or Kubernetes distributions.
+* only cause insignificant load to apiservers. For example, if the master has
+  10GB memory, the migration tool should generate less than 10 qps of single
+  object operations(TODO: measure the memory consumption of PUT operations;
+  study how well the default 10 Mbps bandwidth limit in the oc command work).
+* work for big clusters that have ~10^6 instances of some resource types.
+* make progress in flaky environment, e.g., flaky apiservers, or the migration
+  process get preempted.
+* allow system administrators to track the migration progress.
+
+As to the deliverables,
+* in the short term, providing system administrators with a tool to migrate
+  the API objects to the proper storage versions.
+* in the long term, automating the migration of Kubernetes built-in APIs, CR,
+  aggregated APIs without further burdening system administrators or Kubernetes
+  distributions.
+
+## Proposal
+
+### Alpha workflow
+
+We will create a **kube-migrator** command line tool. At the alpha stage,
```

Comment:
I would greatly prefer CRDs or yamls that can be controlled by kubectl. Why would we add a new command line tool for this? The whole point of our control plane model is that it's extensible.

Reply:
> I would greatly prefer CRDs or yamls that can be controlled by kubectl. Why would we add a new command line tool for this? The whole point of our control plane model is that it's extensible.

This is the basic tool we build up.  This tool would be what is later trigger automatically.  Doing it like this allows us to build up a series of useful pieces.  Nothing precludes adding controller logic around it (this doc describes exactly that), but a standalone CLI tool has value in and of itself.

Reply:
I think Daniel also supports building up the migrator in separable pieces. His concrete [suggestion] (https://github.com/kubernetes/community/pull/2524#discussion_r221402698) was moving the API discovery step to a separate job, that would leave very little logic in the CLI.

I think we should still make the CLI, it will be a thin layer that launches jobs and creates CRs. Users should be able do all the operations via kubectl. The value of the CLI is saving users from editing yamls.

Reply:
Reading the KEP again, I think both the "kube-migrator install" and "kube-migrator progress" commands can be replaced with kubectl one-liners. So I removed the standalone CLI.